Miscellaneous/Best practices for error handling and crash reporting.md

```diff
@@ -0,0 +1,434 @@
+
+- How likely is failure?
+
+- Is ignoring the failure leading to undefined behavior and even worse consequences such as data loss?
+
+- Have you ensured preconditions to be true in other parts of the code?
+
+- Is there a reasonable default value you can provide in case of failure
+
+- Would showing an error message to a user be better than to crash?
+
+Some good resources on this topic include:
+
+- <https://en.wikipedia.org/wiki/Fail-fast_system>
+
+- <https://en.wikipedia.org/wiki/Offensive_programming>
+
+- <https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html>
+
+- <https://www.martinfowler.com/ieeeSoftware/failFast.pdf>
+
+
+### Additional crash logs
+
+Seeing stack trace can very often lead to the clear root cause and conditions under which it happens. Unfortunately, for some of the crashes, this is not the case and any associated information might be useful to investigate, understand, reproduce, and fix the crash.
+
+We can abstract this idea into the following protocol:
+
+```
+public protocol CrashInfoLogger: Sendable {
+    func log(_ message: String)
+}
+```
+And we can implement Firebase crash info logging:
+
+```
+final class FirebaseCrashLogger: CrashLogger {
+    func log(_ message: String) {
+        Logger.app.log("\(message)")
+        Crashlytics.crashlytics().log(message)
+    }
+}
+```
+
+And finally, we can provide a convenient extension for it:
+
+```
+public extension Logger {
+    static let crashInfo: CrashLogger = FirebaseCrashLogger()
+}
+```
+
+This way, we can log some key events during the execution of the program. If the app crashes, these key events will be associated with the crash report in Firebase.
+
+
+## Following trends
+
+Whether it is crashes or errors, some problems might never be reproducible in your local setup. For these issues, we often need to experiment with the fix.
+
+But we need some way to confirm if our fix helped or not. Firebase can provide valuable information about that. The following graph is an example of an issue that was fixed and confirmed by production data:
+
+![](/img/trend.png)
+
+Firebase also provides a lot of valuable associated data. For example, this crash is only happening on iOS 15:
+
+![](/img/ios-15-crashes.png)
+
+This is key information that can help us reproduce and fix the issue.
+
+App Store Connect can also provide valuable information about your crash rate and how you compare to other apps in your category. It is important to keep an eye on the trend and make sure it is not growing.
+
+
+## Legal considerations
+
+Logging errors, or even crashes to third-party systems can have a legal impact on the product you are making. Always check with your legal department what you are allowed to do.
+
+
+## Being proactive
+
+Even if we follow all best practices and invest significant resources into this area, errors and crashes are almost impossible to avoid.
```

Comment:
I'd leave out the almost, I don't think you can ever avoid them 😅 
something always slips by, and it might not be on your end 😂 
```suggestion
Even if we follow all best practices and invest significant resources into this area, errors and crashes are impossible to avoid.
```

Reply:
I was thinking how to formulate this, but I try to be careful with strong words such as "impossible". I guess there might be some super simple app that never crashes?

Reply:
Yep, you're right. Just ignore this comment then 😄